src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4469,9 +4565,64 @@ protected override bool IntersectWith(ref LocalState self, ref LocalState other)
 
                 for (int slot = 1; slot < self.Capacity; slot++)
                 {
-                    bool? selfSlotIsNotNull = self[slot];
-                    bool? intersection = selfSlotIsNotNull & other[slot];
-                    if (selfSlotIsNotNull != intersection)
+                    NullableAnnotation selfAnnotation = self[slot];
+                    NullableAnnotation otherAnnotation = other[slot];
+                    NullableAnnotation intersection;
+
+                    if (selfAnnotation.IsAnyNullable() || otherAnnotation.IsAnyNullable())
+                    {
+                        intersection = selfAnnotation == NullableAnnotation.Nullable || otherAnnotation == NullableAnnotation.Nullable ?
+                                    NullableAnnotation.Nullable : NullableAnnotation.NullableBasedOnAnalysis;
+                    }
+                    else if (selfAnnotation == NullableAnnotation.Unknown)
+                    {
+                        if (otherAnnotation == NullableAnnotation.Unknown || otherAnnotation == NullableAnnotation.NotNullableBasedOnAnalysis)
+                        {
+                            intersection = NullableAnnotation.Unknown;
+                        }
+                        else
+                        {
+                            Debug.Assert(otherAnnotation == NullableAnnotation.NotNullable);
+                            Symbol symbol = variableBySlot[slot].Symbol;
+
+                            if ((object)symbol != null && VariableType(symbol).TypeSymbol?.IsPossiblyNullableReferenceTypeTypeParameter() == true)
```

Comment:
Please consider extracting a local function, `isPossiblyNullableReferenceTypeTypeParameter(slot)`, for use here and below.

Reply:
> Please consider extracting a local function, isPossiblyNullableReferenceTypeTypeParameter(slot), for use here and below.

Coming in the next commit.

---
In reply to: [231339964](https://github.com/dotnet/roslyn/pull/30913#discussion_r231339964) [](ancestors = 231339964)